frontends/p4/sideEffects.h

```diff
@@ -75,39 +87,41 @@ class SideEffects : public Inspector {
     }
 };
 
-// convert expresions so that each expression contains at most one side-effect
-// left-values are converted to contain no side-effects.  An important consequence
-// of this pass is that it converts function calls so that no two parameters
-// (one of which is output) can alias.  This makes the job of the inliner simpler.
-//
-// a[f(x)] = b
-// is converted to
-// tmp = f(x); a[tmp] = b
-//
-// m(n()) is converted to
-// tmp = n()
-// m(tmp)
-//
-// This is especially tricky for handling out and inout arguments.
-// Consider bit f(inout T a, in T b); T g(inout T c);
-// The expression a[g(w)].x = f(a[1].x, g(a[1].x)); is translated as
-// T tmp1 = g(w);           // modifies w
-// T tmp2 = a[1].x;         // save a[1].x
-// T tmp3 = g(a[1].x);      // modifies a[1].x
-// T tmp4 = f(tmp2, tmp3);  // modifies tmp2
-// a[1].x = tmp2;           // copy tmp2 to a[1].x - out argument
-// a[tmp1].x = tmp4;        // assign result of call of f to actual left value
-// Function call proceeds as follows:
-// - arguments are evaluated in order
-// - inout and out arguments produce left-values - these are "saved"
-// - function is called with temporaries for all arguments
-// - out and inout temporaries are copied to the saved left-values in order
-// An assignment statement
-// e = e1;
-// is treated as if it is a call to a function "set(out T v, in T v)"
-// This pass should be run after removing table parameters.
-// FIXME: does not handle select labels
+/** @brief Convert expressions so that each expression contains at most one
+ * side effect.
+ *
+ * Left-values are converted to contain no side-effects.  An important consequence
+ * of this pass is that it converts function calls so that no two parameters
+ * (one of which is output) can alias.  This makes the job of the inliner simpler.
+ * For example:
+ * - The expression ```a[f(x)] = b;``` is converted to ```tmp = f(x); a[tmp] = b;```
+ * - And ```m(n())``` is converted to ```tmp = n(); m(tmp)```
+ *
+ * This is especially tricky for handling out and inout arguments.  Consider
+ * two function prototypes, ```bit f(inout T a, in T b); T g(inout T c);```.
+ * The expression ```a[g(w)].x = f(a[1].x, g(a[1].x));``` is translated as
+ *
+```
+T tmp1 = g(w);           // modifies w
+T tmp2 = a[1].x;         // save a[1].x
+T tmp3 = g(a[1].x);      // modifies a[1].x
+T tmp4 = f(tmp2, tmp3);  // modifies tmp2
+a[1].x = tmp2;           // copy tmp2 to a[1].x - out argument
+a[tmp1].x = tmp4;        // assign result of call of f to actual left value
+```
+ *
+ * Translating function calls proceeds as follows:
+ * - arguments are evaluated in order
+ * - inout and out arguments produce left-values---these are "saved"
+ * - function is called with temporaries for all arguments
+ * - out and inout temporaries are copied to the saved left-values in order
+ *
+ * An assignment statement ```e = e1;``` is treated as if it is a call to a
```

Comment:
I was too pedantic, this just means that the LHS is evaluated first.

For such a simple language, P4 has a really nasty semantics due to side-effects and out parameter. This pass was one of the hardest to write. After it has been run everything is much simpler.